src/containers/Header/NotificationsTray/index.jsx

```diff
@@ -1,28 +1,54 @@
 import styles from './style.postcss';
 
 import React, { PureComponent } from 'react';
-import classnames from 'classnames';
+import PropTypes from 'prop-types';
 import Cursor from 'immutable-cursor';
 import { CSSTransitionGroup } from 'react-transition-group';
+import classnames from 'classnames';
+import { fetch, buildUrl } from 'domain/Api';
 import Config from 'domain/Config';
+import createApiActionCreator from 'domain/createApiActionCreator';
 import connect from 'domain/connect';
 import Icon from 'components/Icon';
 import Callout from 'components/Callout';
+import Button from 'components/Button';
+import Checkbox from 'components/Checkbox';
 
 class NotificationsTray extends PureComponent {
   constructor() {
     super();
-    this.state = { viewingNotification: null };
+    this.state = {
+      viewingNotification: null,
+      isMarkingMode: false,
+      notificationIdsToMarkRead: [],
+    };
     this._onNotificationClicked = this._onNotificationClicked.bind(this);
     this._onClickBackToNotifications = this._onClickBackToNotifications.bind(this);
     this._onCalloutOpenStateChanged = this._onCalloutOpenStateChanged.bind(this);
     this._renderCalloutPopupContent = this._renderCalloutPopupContent.bind(this);
   }
 
   _onNotificationClicked(ev) {
+    ev.preventDefault();
+    ev.stopPropagation();
     const element = ev.currentTarget;
     const { notificationId } = element.dataset;
-    this.setState({ viewingNotification: this.props.notifications.get(notificationId) });
+
+    if (this.state.isMarkingMode) {
+      if (! this.state.notificationIdsToMarkRead.includes(notificationId)) {
+        const notificationIds = [...this.state.notificationIdsToMarkRead, notificationId];
+        return this.setState({
+          notificationIdsToMarkRead: notificationIds,
+        });
+      }
+      const notificationIds = [...this.state.notificationIdsToMarkRead.filter(
```

Comment:
A `Set` is better for this. `set.has(value)`, `set.add(value)`, etc. Values in the set are always unique.